saleor/tax/migrations/0004_migrate_vatlayer.py

```diff
@@ -0,0 +1,187 @@
+    class.
+    """
+
+    Product = apps.get_model("product", "Product")
+    ProductType = apps.get_model("product", "ProductType")
+    TaxClass = apps.get_model("tax", "TaxClass")
+
+    product_types = (
+        ProductType.objects.filter(metadata__has_key=VATLAYER_CODE_META_KEY)
+        .values("id", "metadata")
+        .order_by("pk")
+    )
+    for batch_pks in queryset_in_batches(product_types):
+        tax_classes_from_product_types = defaultdict(list)
+        product_types = ProductType.objects.filter(pk__in=batch_pks)
+        for product_type in product_types:
+            tax_class_name = product_type.metadata.get(VATLAYER_CODE_META_KEY)
+            tax_classes_from_product_types[tax_class_name].append(product_type.pk)
+
+        for name, ids in tax_classes_from_product_types.items():
+            tax_class, _ = TaxClass.objects.get_or_create(name=name)
+            ProductType.objects.filter(id__in=ids).update(tax_class=tax_class)
+
+    products = (
+        Product.objects.filter(metadata__has_key=VATLAYER_CODE_META_KEY)
+        .values("id", "metadata")
+        .order_by("pk")
+    )
+    tax_classes_from_products = defaultdict(list)
+    for batch_pks in queryset_in_batches(products):
+        products = Product.objects.filter(pk__in=batch_pks)
+        for product in products:
+            tax_class_name = product.metadata.get(VATLAYER_CODE_META_KEY)
+            tax_classes_from_products[tax_class_name].append(product.pk)
+
+        for name, ids in tax_classes_from_products.items():
+            tax_class, _ = TaxClass.objects.get_or_create(name=name)
+            Product.objects.filter(id__in=ids).update(tax_class=tax_class)
+
+
+def create_tax_configurations(apps, vatlayer_configs):
+    TaxConfiguration = apps.get_model("tax", "TaxConfiguration")
+    TaxConfigurationPerCountry = apps.get_model("tax", "TaxConfigurationPerCountry")
+
+    # Map of countries to use origin country's tax, based on the
+    # `countries_to_calculate_taxes_from_origin` setting. If a country code appears
+    # more than once in the list, we override is with the last seen origin_country.
+    use_origin_country_map = {}
+
+    for vatlayer_config in vatlayer_configs:
+        config_dict = {
+            item["name"]: item["value"] for item in vatlayer_config.configuration
+        }
+        channel = vatlayer_config.channel
+
+        # Migrate `countries_to_calculate_taxes_from_origin` setting.
+        origin_country = _clear_country_code(config_dict["origin_country"])
+        countries_to_calculate_taxes_from_origin = _clear_str_list_country_codes(
+            config_dict["countries_to_calculate_taxes_from_origin"]
+        )
+        for country in countries_to_calculate_taxes_from_origin:
+            use_origin_country_map[country] = origin_country
+
+        # Migrate `excluded_countries` to new tax configuration.
+        excluded_countries = _clear_str_list_country_codes(
+            config_dict["excluded_countries"]
```

Comment:
Assuming that you already check that, but I would propose to double check if the `config_dict` have always added fields like `excluded_countries`

Reply:
Especially when you're migrating from version like 2.11 where you had active vatlayer (this version didn't have fields like `excluded_countries`).

Reply:
Good point.